src/cdk/stepper/stepper.ts

```diff
@@ -0,0 +1,169 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  ContentChildren,
+  EventEmitter,
+  Input,
+  Output,
+  QueryList,
+  Directive,
+  ViewChildren,
+  // This import is only used to define a generic type. The current TypeScript version incorrectly
+  // considers such imports as unused (https://github.com/Microsoft/TypeScript/issues/14953)
+  // tslint:disable-next-line:no-unused-variable
+  ElementRef,
+  Component,
+  ContentChild,
+  ViewChild,
+  TemplateRef
+} from '@angular/core';
+import {LEFT_ARROW, RIGHT_ARROW, ENTER, SPACE} from '../keyboard/keycodes';
+import {CdkStepLabel} from './step-label';
+
+/** Used to generate unique ID for each stepper component. */
+let nextId = 0;
+
+/** Change event emitted on selection changes. */
+export class CdkStepperSelectionEvent {
+  /** The index of the step that is newly selected during this change event. */
+  selectedIndex: number;
+
+  /** The index of the step that was previously selected. */
+  previouslySelectedIndex: number;
+
+  /** The new step component that is selected ruing this change event. */
+  selectedStep: CdkStep;
+
+  /** The step component that was previously selected. */
+  previouslySelectedStep: CdkStep;
+}
+
+@Component({
+  selector: 'cdk-step',
+  templateUrl: 'step.html',
+})
+export class CdkStep {
+  /** Template for step label if it exists. */
+  @ContentChild(CdkStepLabel) stepLabel: CdkStepLabel;
+
+  /** Template for step content. */
+  @ViewChild(TemplateRef) content: TemplateRef<any>;
+
+  /** Label of the step. */
+  @Input()
+  label: string;
+
+  constructor(private _stepper: CdkStepper) { }
+
+  /** Selects this step component. */
+  select(): void {
+    this._stepper.selected = this;
+  }
+}
+
+@Directive({
+  selector: 'cdk-stepper',
+  host: {
+    '(focus)': '_setStepFocused()',
+    '(keydown)': '_onKeydown($event)',
+  },
+})
+export class CdkStepper {
+  /** The list of step components that the stepper is holding. */
+  @ContentChildren(CdkStep) _steps: QueryList<CdkStep>;
+
+  /** The list of step headers of the steps in the stepper. */
+  _stepHeader: QueryList<ElementRef>;
+
+  /** The index of the selected step. */
+  get selectedIndex() { return this._selectedIndex; }
+  set selectedIndex(index: number) {
+    if (this._selectedIndex != index) {
+      this._emitStepperSelectionEvent(index);
+      this._setStepFocused(this._selectedIndex);
+    }
+  }
+  private _selectedIndex: number = 0;
+
+  /** Returns the step that is selected. */
+  get selected() { return this._steps[this.selectedIndex]; }
+  /** Sets selectedIndex as the index of the provided step. */
```

Comment:
nit: can just put 1 comment for set & get since they represent the same property